core/src/main/java/io/questdb/griffin/SqlOptimiser.java

```diff
@@ -3801,6 +3801,126 @@ private ExpressionNode rewriteAggregate(ExpressionNode agg, QueryModel model) {
+             Type 1 Optimisations: Will be optimised by adding order by clause and changing model type
+             Type 2 Optimisations: Will be done only by changing model type to prevent invoking group by workers
+             */
+            int optimisationType = 0;
+            if (rhs != null && ast.type == 8 && Chars.equals(designatedTimestampColumn, rhs)) {
+                if (Chars.equalsIgnoreCase(token, "LAST") || Chars.equalsIgnoreCase(token, "MAX"))
+                    optimisationType = 1;
+                else if (Chars.equalsIgnoreCase(token, "FIRST") || Chars.equalsIgnoreCase(token, "MIN"))
+                    optimisationType = 2;
+            }
+
+            /**
+             Core logic for issue #4231 will be applicable under two conditions
+             Condition1: QueryColumn contains FIRST/LAST/max/min function keyword
+             Condition2: QueryColumn for LAST/max keyword should be same as designatedTimestamp column
+             Core logic for changing query plan
+             - Add limit 1 via expressionNode to parent model
+             - Change model to select-choose from select-group-by
+             - Column alias by default is LAST/max if alias is not provided, replace with original column name
+             - Add order by clause to nested model
+             Call this recursively for nested models
+             */
+            if (optimisationType == 1 || optimisationType == 2) {
+                QueryModel newNestedModel = queryModelPool.next();
+                ExpressionNode lowerLimitNode = expressionNodePool.next();
+                lowerLimitNode.token = "1";
+                lowerLimitNode.type = 2;
+                parent.setLimit(lowerLimitNode, null);
+
+                //change model type to select-choose
+                parent.setSelectModelType(1);
+
+                //change ast params
+                ast.token = rhs;
+                ast.paramCount = 0;
+                ast.type = 4;
+
+                ExpressionNode newTimestampNode = expressionNodePool.next();
+                newTimestampNode.token = designatedTimestampColumn.toString();
+                if (optimisationType == 1)
+                    newNestedModel.addOrderBy(newTimestampNode, QueryModel.ORDER_DIRECTION_DESCENDING);
+                ObjList<QueryModel> joinModels = parent.getNestedModel().getJoinModels();
+                for (int i = 1, n = joinModels.size(); i < n; i++) {
+                    newNestedModel.addJoinModel(joinModels.getQuick(i));
+                }
+                newNestedModel.setTableNameExpr(nestedModel.getTableNameExpr());
+                newNestedModel.setModelType(nestedModel.getModelType());
+                newNestedModel.setTimestamp(nestedModel.getTimestamp());
+                newNestedModel.copyColumnsFrom(nestedModel, queryColumnPool, expressionNodePool);
+                parent.setNestedModel(newNestedModel);
+                modifiedNestedModel = newNestedModel;
+            }
```

Comment:
If the rewrite took place, we should stop the recursion - similarly to how it's done in `rewriteCountDistinct()`.

Reply:
@puzpuzpuz in that case we have to optimise union models before base models, consider a case given like
https://github.com/questdb/questdb/pull/4676/files#diff-096eed90f4de10911f6a8549ad3026cc552b8495e032e785b33089ec15c4e2cdR518
If we optimise union model after base model, then only last union will be optimised as rest of the parent models won't satisfy condition of nestedModel.getUnionModel() == null , however if we remove this condition as well, the execution will return only after optimising the first union model, not the rest

Reply:
We should optimize only the most nested models that involve a "physical" table. There is no point in transforming `select last(ts) from ((tab1 union tab2 order by ts) timestamp(ts))` to `select ts from ((tab1 union tab2 order by ts) timestamp(ts)) limit -1`. Instead, we should transform subqueries, if they meet the criteria. That's because we can do backward scan only for "physical" tables.

So, the rewrite should look for all variations of `select last(ts) from tab` (also, first/min/max) query in the nested queries, rewrite them, and call it a day.

Reply:
@puzpuzpuz current implementation doesn't optimise : 
`select LAST(ts) from (ticker_price_2 union ticker_price_3 order by ts) timestamp(ts)`
The rewritten model for such a query looks like this:
 ```
select-group-by LAST(ts) ts1 from (select-choose [ts] ticker, ts from 
 (select-choose [ticker, ts] ticker, ts from (select [ticker, ts] from 
 ticker_price_2 timestamp (ts)) union select-choose [ticker, ts] ticker, ts from 
 (select [ticker, ts] from ticker_price_3 timestamp (ts))) order by ts) timestamp (ts)
```

The query which I pointed out :
`select FIRST(ts) from y union select LAST(ts) from y union select min(ts) from y  union select max(ts) from y`
has 4 variations of `select last(ts) from tab`, so if we **optimise union models after base model(considering we return after optmising base model)**, it will not optimise 

1.  `select LAST(ts) from y`
2. `select min(ts) from y`
3. `select max(ts) from y`
only optimised model will be` select FIRST(ts) from y`

that's why optimising union models first will optimise models in this order:
1. `select max(ts) from y`
2. `select min(ts) from y`
3. `select LAST(ts) from y`
4. `select FIRST(ts) from y`

I'll check the case with explicit where clause

Reply:
> if we **optimise union models after base model(considering we return after optmising base model)**, it will not optimise

I see, thanks for the clarification. It makes sense.

> It already does backward scans , do you want me to handle any other cases?

Not really - unless you have some other cases on your mind. But it's certainly a good idea to add more tests, e.g. around rewrite for a query with a filter.

Reply:
@puzpuzpuz I have one more doubt here : 
**(ts)** is designated timestamp column
Query 1 - `select LAST(ts) from ticker_price where ticker = 'S3'`
Query 2 -  `SELECT LAST(ts) from (select * from ticker_price where ticker = 'S3')`
Both of these queries **should have** same query planner and answers, but current executor is doing **"backward scan" in **Query 1**** and **"forward scan" in **Query 2**** because of current logic implementation
It should be backward scan for both of the cases right?

Reply:
> Does Query plan for Query 2 need to be checked (it's a very specific case)?

No, let's keep things simple, so let's target Query 1.

As for Query 2, it could be rewritten to Query 1 as a separate step in the SqlOptimizer, so that rewrites can kick in later. But that's another story.